AX: 8 properties are unnecessarily cached for every single AXIsolatedObject

twilco opened the pull request:
#### 3cc438654ebc758e803dfa01b3ab638ace726a96
<pre>
AX: 8 properties are unnecessarily cached for every single AXIsolatedObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=290585">https://bugs.webkit.org/show_bug.cgi?id=290585</a>
<a href="https://rdar.apple.com/148061952">rdar://148061952</a>

Reviewed by Chris Fleizach.

Eight properties were unnecessarily cached for every isolated object, and are no longer cached except in the rare case
they are non-default.

AXProperty::ColorValue was always cached because our &quot;check for default value&quot; logic in AXIsolatedObject::setProperty
compared against Color(), which is transparent black, but AccessibilityNodeObject::colorValue returns Color::black
for anything that isn&apos;t a color input, and these are not equal. This is fixed by some special-case handling when
caching and returning AXProperty::ColorValue.

Furthermore, our is-default check for `Color`s is generally much smarter and caches far fewer values, e.g. for properties
like AXProperty::BackgroundColor and AXProperty::LinethroughColor. Prior to this commit, we compare to Color() without
converting to RGBA, which is a lossy conversion. This meant that the colors would be effectively the same as the default,
but still get cached. Now we eagerly convert to SRGBA before comparing colors.

AXProperty::{InvalidStatus, AutoCompleteValue, PopupValue, LiveRegionRelevant} all fell back to returning a non-empty
string in the default case when the corresponding attribute wasn&apos;t present, meaning we always cached something for these
properties. With this commit, we now only cache something when a value is explicitly set.

AXProperty::SpeechHint has been changed to AXProperty::SpeakAs. This propery was similar to the previous group - we always
computed a non-empty string, and thus always cached something.

AXProperty::StringValue used to cache null strings, but not empty ones, as an optimization. Prior to ENABLE(AX_THREAD_TEXT_APIS),
we needed this distinction to know whether we should fallback to using the cached AXProperty::AttributedString as the string value.
However, with ENABLE(AX_THREAD_TEXT_APIS), we now compute attributed strings on-the-fly, and thus don&apos;t need to avoid
considering null string values to be default, allowing us to avoid caching this property ~95% more often (based on data from gmail.com).

AXProperty::SortDirection always used to be cached because our is-default check looked for AccessibilitySortDirection::None,
but for anything that column header or row header, we would return AccessibilitySortDirection::Invalid. This commit addresses
the issue by making Invalid the first variant in the enum, meaning it will be the default, since we cast this enum to an
int before caching it.

This saves roughly ~218 mb of memory on <a href="http://html.spec.whatwg.org.">http://html.spec.whatwg.org.</a>

(567769 objects * 8 fewer properties per object * 48 bytes (sizeof(AXPropertyValueVariant)))

The savings are probably even better considering the significantly smarter Color default checks.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::autoCompleteValue const):
(WebCore::AXCoreObject::invalidStatus const):
(WebCore::AXCoreObject::popupValue const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::liveRegionRelevant const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::liveRegionRelevant const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::speakAs const):
(WebCore::AccessibilityObject::explicitAutoCompleteValue const):
(WebCore::AccessibilityObject::explicitInvalidStatus const):
(WebCore::AccessibilityObject::explicitPopupValue const):
(WebCore::AccessibilityObject::speakAsProperty const): Deleted.
(WebCore::AccessibilityObject::autoCompleteValue const): Deleted.
(WebCore::AccessibilityObject::invalidStatus const): Deleted.
(WebCore::AccessibilityObject::popupValue const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::speechHint const):
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::AccessibilityObject::speechHintAttributeValue const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::colorValue const):
(WebCore::AXIsolatedObject::optionSetAttributeValue const):
(WebCore::AXIsolatedObject::stringValue const):
(WebCore::AXIsolatedObject::ancestorFlags const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(-[WebAccessibilityObjectWrapperBase baseAccessibilitySpeechHint]):

Canonical link: <a href="https://commits.webkit.org/292834@main">https://commits.webkit.org/292834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c1f7f9399ae8cb0bcd3882e6f481557d8dc81f7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97160 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16783 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102241 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47685 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99205 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17076 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25233 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74010 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31219 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100163 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12892 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87870 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54356 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12645 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5742 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/47234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82699 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5819 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104263 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24235 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17671 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83056 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/24611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83993 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82466 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20767 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4702 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17739 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24199 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29350 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24022 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27334 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25595 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -392,6 +392,19 @@ AXCoreObject* AXCoreObject::nextSiblingIncludingIgnoredOrParent() const
 
+String AXCoreObject::autoCompleteValue() const
+{
+    String explicitValue = explicitAutoCompleteValue();
```
can this be auto explicitvalue ?

twilco replied:
I think it's kind of nice to know that it's a `String`, since the function name doesn't make it super clear (without looking into the function definition, "autocomplete value" could be a string, or an enum matching one of the pre-defined `aria-autocomplete` values, or something else — but the explicit type answers this question without any further investigation)

twilco committed:
AX: 8 properties are unnecessarily cached for every single AXIsolatedObject
https://bugs.webkit.org/show_bug.cgi?id=290585
rdar://148061952

Reviewed by Chris Fleizach.

Eight properties were unnecessarily cached for every isolated object, and are no longer cached except in the rare case
they are non-default.

AXProperty::ColorValue was always cached because our "check for default value" logic in AXIsolatedObject::setProperty
compared against Color(), which is transparent black, but AccessibilityNodeObject::colorValue returns Color::black
for anything that isn't a color input, and these are not equal. This is fixed by some special-case handling when
caching and returning AXProperty::ColorValue.

Furthermore, our is-default check for `Color`s is generally much smarter and caches far fewer values, e.g. for properties
like AXProperty::BackgroundColor and AXProperty::LinethroughColor. Prior to this commit, we compare to Color() without
converting to RGBA, which is a lossy conversion. This meant that the colors would be effectively the same as the default,
but still get cached. Now we eagerly convert to SRGBA before comparing colors.

AXProperty::{InvalidStatus, AutoCompleteValue, PopupValue, LiveRegionRelevant} all fell back to returning a non-empty
string in the default case when the corresponding attribute wasn't present, meaning we always cached something for these
properties. With this commit, we now only cache something when a value is explicitly set.

AXProperty::SpeechHint has been changed to AXProperty::SpeakAs. This propery was similar to the previous group - we always
computed a non-empty string, and thus always cached something.

AXProperty::StringValue used to cache null strings, but not empty ones, as an optimization. Prior to ENABLE(AX_THREAD_TEXT_APIS),
we needed this distinction to know whether we should fallback to using the cached AXProperty::AttributedString as the string value.
However, with ENABLE(AX_THREAD_TEXT_APIS), we now compute attributed strings on-the-fly, and thus don't need to avoid
considering null string values to be default, allowing us to avoid caching this property ~95% more often (based on data from gmail.com).

AXProperty::SortDirection always used to be cached because our is-default check looked for AccessibilitySortDirection::None,
but for anything that column header or row header, we would return AccessibilitySortDirection::Invalid. This commit addresses
the issue by making Invalid the first variant in the enum, meaning it will be the default, since we cast this enum to an
int before caching it.

This saves roughly ~218 mb of memory on http://html.spec.whatwg.org.

(567769 objects * 8 fewer properties per object * 48 bytes (sizeof(AXPropertyValueVariant)))

The savings are probably even better considering the significantly smarter Color default checks.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::autoCompleteValue const):
(WebCore::AXCoreObject::invalidStatus const):
(WebCore::AXCoreObject::popupValue const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::liveRegionRelevant const):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::liveRegionRelevant const): Deleted.
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::speakAs const):
(WebCore::AccessibilityObject::explicitAutoCompleteValue const):
(WebCore::AccessibilityObject::explicitInvalidStatus const):
(WebCore::AccessibilityObject::explicitPopupValue const):
(WebCore::AccessibilityObject::speakAsProperty const): Deleted.
(WebCore::AccessibilityObject::autoCompleteValue const): Deleted.
(WebCore::AccessibilityObject::invalidStatus const): Deleted.
(WebCore::AccessibilityObject::popupValue const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::speechHint const):
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::AccessibilityObject::speechHintAttributeValue const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::colorValue const):
(WebCore::AXIsolatedObject::optionSetAttributeValue const):
(WebCore::AXIsolatedObject::stringValue const):
(WebCore::AXIsolatedObject::ancestorFlags const): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(-[WebAccessibilityObjectWrapperBase baseAccessibilitySpeechHint]):

Canonical link: https://commits.webkit.org/292834@main
